modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,167 @@
+import * as NodeType from '../dom/NodeType';
+import { FormatVars } from './FormatTypes';
 import * as FormatUtils from './FormatUtils';
 import * as MatchFormat from './MatchFormat';
 
 export type FormatChangeCallback = (state: boolean, data: { node: Node; format: string; parents: any }) => void;
 
-export type FormatCallbacks = Record<string, FormatChangeCallback[]>;
+export type RegisteredFormats = Record<string, CallbackGroup>;
 
-export interface FormatData {
-  similar?: boolean;
-  callbacks: FormatChangeCallback[];
+export interface UnbindFormatChanged {
+  unbind: () => void;
 }
 
-export type RegisteredFormats = Record<string, FormatData>;
+interface CustomCallback {
+  readonly state: Cell<boolean>;
+  readonly similar: boolean;
+  readonly vars: FormatVars;
+  readonly callback: FormatChangeCallback;
+}
 
-export interface UnbindFormatChanged {
-  unbind: () => void;
+interface CallbackGroup {
+  // Bundle all callbacks with similar=true and unspecified vars
+  readonly similarState: Cell<boolean>;
+  readonly similarCallbacks: FormatChangeCallback[];
+  // Also bundle all callbacks with similar=false and unspecified vars
+  readonly similarFalseState: Cell<boolean>;
+  readonly similarFalseCallbacks: FormatChangeCallback[];
+  // Then for callbacks that have variables, handle everything separately
+  readonly others: CustomCallback[];
 }
 
 const setup = (registeredFormatListeners: Cell<RegisteredFormats>, editor: Editor) => {
-  const currentFormats = Cell<Record<string, FormatChangeCallback[]>>({ });
-
   registeredFormatListeners.set({});
 
   editor.on('NodeChange', (e) => {
-    updateAndFireChangeCallbacks(editor, e.element, currentFormats, registeredFormatListeners.get());
+    updateAndFireChangeCallbacks(editor, e.element, registeredFormatListeners.get());
   });
-};
-
-const updateAndFireChangeCallbacks = (editor: Editor, elm: Element, currentFormats: Cell<FormatCallbacks>, formatChangeData: RegisteredFormats) => {
-  const formatsList = Obj.keys(currentFormats.get());
-  const newFormats: FormatCallbacks = { };
-  const matchedFormats: FormatCallbacks = { };
-
-  // Ignore bogus nodes like the <a> tag created by moveStart()
-  const parents = Arr.filter(FormatUtils.getParents(editor.dom, elm), (node) => node.nodeType === 1 && !node.getAttribute('data-mce-bogus'));
-
-  // Check for new formats
-  Obj.each(formatChangeData, (data: FormatData, format: string) => {
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
 
-        matchedFormats[format] = data.callbacks;
-        return false;
-      }
+  editor.on('FormatApply FormatRemove', (e: EditorEvent<FormatEvent>) => {
```

Comment:
This logic is incorrect... it won't work when you pass in `node` to apply/remove, which is also why I had said in the call the other day that needs to be passed through.

- [ ] Fix this to take into account the node/range like object passed to the relevant format functions